### ISSUE TITLE
`.findAll()` etc return instance of `Collection` class?

### ISSUE THREAD

**Issue by overlookmotel** [sequelize/sequelize#6654]
Subclassing `Array`is now possible in ES6. So, I wonder if it'd be good for Sequelize methods that return a collection of model instances, instead of returning a plain array, to return an instance of a custom class?

## How it could work

```js
Sequelize.Collection extends Array { ... }
Model.Collection extends Sequelize.Collection { ... }
// NB `Model` in line above is the class for an *individual* Model
```

e.g.:

```js
const tasks = await Task.findAll();
// tasks instanceof Task.Collection == true
// tasks instanceof Sequelize.Collection == true
// tasks instanceof Array == true
```

So `Task.findAll()` returns an instance of `Task.Collection` rather than a plain `Array`. But because `Task.Collection` has `Array` at the root of it's prototype chain, all usual Array methods would still work (see proviso about Node v4 below).

### 1. Operate on a whole collection

This would allow new methods that operate on a collection. e.g.:

```js
const tasks = await Task.findAll( { where: { status: 'In progress' } } );
await tasks.update( { status: 'Done' } );
```

`tasks.update()` is equivalent to:

```js
// Multiple queries
await Promise.all( tasks.map( task => {
    return task.update( { status: 'Done' } );
} );

// Better as creates only one database query
await Task.update( { status: 'Done' }, { where: { id: tasks.map( task => task.id ) } );
```

Another example:

```js
const tasks = await Task.findAll();

// ... then later in code ...
await tasks.getUser();
```

`tasks.getUser()` is equivalent to:

```js
// Multiple queries
await Promise.all( tasks.map( task => {
    return task.getUser().then( user => {
        task.user = user;
    } );
} );

// Single query
const userIds = tasks.map( task => task.userId );
const users = await Users.findAll( { where: { id: userIds } } );
for (let task of tasks) {
    task.user = _.find( users, user => user.id = task.userId );
}
// NB could be made more efficient by removing duplicates from `userIds`,
// getting users in order of `id`, and implementing a binary search in place of `_.find()`
```

In all the above use cases, it's pure sugar. But in the last case, if the implementation was made as efficient as possible, it's sugar for quite a complex set of operations.

### 2. Pub-sub

This is way off on the horizon, but I've been thinking about at some point trying to implement some form of pub-sub for Sequelize.

For this, it'd be useful to be able to register event listeners for a *collection*.

```js
const tasks = await Task.findAll( { where: { status: 'In progress' } } );
tasks.onInsert( (newTask, index) => { ... } );
```

The above isn't really sugar - it's something entirely new, and requires the concept of a collection to make sense.

### Node v6

Node v6 has complete support for subclassing `Array` so this wouldn't break anything.

`Array.isArray( collection )` and `collection instanceof Array` behave as before. All array methods `.map()`, `.push()` etc work as before, and `.length` is updated as it should be.

### Node v4

Node v4 has incomplete support for Array subclassing (see [here](http://node.green/#Array-is-subclassable).

So in Node v4, things like this would break:

```js
const tasks = await Task.findAll();
tasks.map( task => { ... } );
```

### How to support Node v4

I can think of 3 options:

1. Have this as a feature that is switched off by default and you have to enable it with `sequelize.options.collections = true`. If `options.collections == false`, an Array is returned same as it is now.

2. Default to `sequelize.options.collections = true` and users have to switch it off in Node v4 if they want to `.map()` over model collections.

3. As (1), but set default `sequelize.options.collections` to `true` if running Node v6, `false` if Node v4.

I'm not sure that option (3) is a good idea - it'd change the default API depending on what version of Node you're on.

Depending on which of the above options was chosen, it could or could not be a semver major change.

## Conclusion

What do you guys think?

**Comment by felixfbecker** [comment]
I think this is a great idea, I used this already in some projects:
https://github.com/felixfbecker/merkel/blob/master/src/migration.ts#L64

It of course only makes sense when we actually provide methods on it that are useful, like operation on all the items. Ideally, they would just call static model methods, so all operations are also possible if you don't have a collection but know the IDs.
I think these implementations are quite complex and I don't know if he use case is really that great.
No matter what, we should fully support it, not through some flag - typings for example would become a nightmare otherwise. `map()` is a crucial operation. Do you know what exactly does not work? Is it erroring? Maybe we could implement the methods ourself by calling `Array.prototype[method].apply()`. [Node 4 LTS support ends as late as April 2018](https://github.com/nodejs/LTS).

**Comment by overlookmotel** [comment]
@felixfbecker Glad you like it!

My reasoning for bringing this up now, is if it's not going to be disabled by default, it'd be a semver major change so wanted to get it in before Sequelize 4.0.0 if possible. The actual methods could be added later as that wouldn't be breaking changes.

Personally, even without the methods, I'd still find it useful in my sequelize plugins for doing e.g. `result instanceof Sequelize.Collection`.

As for the methods, I find myself writing this kind of stuff (e.g. an equivalent of `tasks.getUser()`) quite a lot in hooks code, so I do think there's a genuine use case.

#### Node v4:

You're right, internally, we could use `Array.prototype.map.apply()` etc to ensure no errors on Node v4.

But the problem is for the end user consuming the result of Sequelize methods. They may well have existing code like `(await Task.findAll()).map( ... )` and that would break on Node v4.

**Comment by felixfbecker** [comment]
If we can make it compatible with Node 4, it would be a semver-minor, as we return a subclass of Array and by Liskov's substitution principle any instance of a parent class can be replaced by an instance of a child class without breaking any code. Therefor this could be added later in the lifecycle of v4 (there are already way to many things on the v4 milestone).

If we cannot make it compatible with Node 4, it is a breaking change that must wait for v5/6/7 until Node 4 is not LTS anymore.

I don't understand your last paragraph, could you explain a bit more?

**Comment by overlookmotel** [comment]
Liskov's principle is right in principle, but in practice, extending "exotic" Javascript native classes like `Array` wasn't supported at all until ES6, and full compliance not achieved until Node v6. I think it's the `.length` attribute that causes the trouble. With an array, when you do `arr[1] = value`, `arr.length` magically changes its value, but with subclasses of Array this behavior wasn't preserved.

Actually, I've looked into this further, and it looks like it may be a non issue.

According to [this](http://node.green/#Array-is-subclassable) subclassing of `Array` is supported as far back as node v4.3.2. The thing that doesn't work until node v6 is array methods like `.map()` called on a subclass of Array return an array, rather than an instance of the subclass.

Here's the failing test in Node v4:

```js
class C extends Array {}
var c = new C();
expect( c.map(Boolean) instanceof C).to.be.true;
```

I don't think this is a deal-breaker. We could either polyfill the offending prototype methods on `Collection` class, or leave it as is. Either way, it's not going to break anyone's code.

**Comment by overlookmotel** [comment]
Only way I can conceive that this is a breaking change is if someone is testing for whether the result of a Sequelize method is a collection or an object by using `result.__proto__ == Array.prototype`. And that would be pretty non-standard (not to say reckless!)

**Comment by overlookmotel** [comment]
@mickhansen @janmeier You got any thoughts on this?

**Comment by felixfbecker** [comment]
@overlookmotel If someone is doing it like this, he has not understood OOP :D

**Comment by overlookmotel** [comment]
@felixfbecker Ah no, collection method still exist and work fine, but the point is that the collection methods just return a plain array rather than an instance of the subclass.

Node v6:

```js
class C extends Array {}
const c = new C();
const mapped = c.map( () => {} );

mapped instanceof C // == true
mapped instanceof Array // == true
mapped.constructor = C // == true
```

Node v4:

```js
class C extends Array {}
const c = new C();
const mapped = c.map( () => {} );

mapped instanceof C // == false
mapped instanceof Array // == true
mapped.constructor = Array // == true
```

**Comment by overlookmotel** [comment]
> @overlookmotel If someone is doing it like this, he has not understood OOP :D

ES6 spec is clear on the matter. But obviously ES6 support has been phased in slowly to V8, so older versions of Node didn't support all of the ES6 spec.

**Comment by felixfbecker** [comment]
That was my point, you would expect another sequelize collection after calling map

**Comment by overlookmotel** [comment]
Well I don't know - it depends what you map the instances to.

```js
const tasks = await Task.findAll( { attributes: ['id'] } );
const taskIds = tasks.map( task => task.id );
```

Wouldn't it make more sense for `taskIds` here to be a plain array? And the proposed new collection methods wouldn't work either e.g. `taskIds.getUser()`.

**Comment by felixfbecker** [comment]
map is a bad example, but you can always do Array.from() to get a plain array. But think about collection.filter().

Btw I'm against a flag to configure this. As the maintainer of the Typescript typings and user of Sequelize with typings, this would make Sequelize impossible to type. There is also no reason for a flag thanks to Liskov.

**Comment by overlookmotel** [comment]
Yes, I agree. I think the differences between Node v4 and Node v6 aren't unsurmountable. The differences could be polyfilled, or left as they are. Either way, I think it's possible to implement this without breaking changes.

However, for safety, I think it'd be ideal if this change happened before Sequelize 4.0.0. But I don't know what current thinking on schedule for that is.

**Comment by overlookmotel** [comment]
Can we leave this issue open please?
